distribution/lib/Standard/Table/0.2.32-SNAPSHOT/src/Data/Matching.enso

```diff
@@ -0,0 +1,165 @@
+     By default, a warning is issued, but the operation proceeds.
+     If set to `Report_Error`, the operation fails with a dataflow error.
+     If set to `Ignore`, the operation proceeds without errors or warnings.
+   - warning_callback: A callback specifying the function to call when issuing a
+     warning. This is a temporary workaround to allow for testing the warning
+     mechanism. Once the proper warning system is implemented, this argument
+     will become obsolete and will be removed. No user code should use this
+     argument, as it will be removed in the future.
+
+   > Example
+     Selects objects matching one of the provided patterns, preserving the input order.
+
+         Matching.match_criteria ["foo", "foobar", "quux", "baz", "Foo"] [".*ba.*", "f.*"] matching_strategy=(Regex case_sensitivity=True) == ["foo", "foobar", "baz"]
+
+   > Example
+     Selects pairs matching their first element with the provided criteria and
+     ordering the result according to the order of criteria that matched them.
+
+         Matching.match_criteria [Pair "foo" 42, Pair "bar" 33, Pair "baz" 10, Pair "foo" 0, Pair 10 10] ["bar", "foo"] reorder=True name_mapper=_.name == [Pair "bar" 33, Pair "foo" 42, Pair "foo" 0]
+match_criteria : Vector Any -> Vector Text -> Boolean -> (Any -> Text) -> Matching_Strategy -> Problem_Behavior -> (Text -> Nothing) -> Vector Any ! No_Matches_Found
+match_criteria objects criteria reorder=False name_mapper=(x->x) matching_strategy=(Exact case_sensitivity=True) on_problems=Report_Warning warning_callback=here.report_warning_placeholder = Panic.recover <|
+    [objects, criteria, reorder, name_mapper, matching_strategy, on_missing, warning_callback] . each Panic.rethrow
+
+    # match_matrix . at i . at j specifies whether objects.at i matches criteria.at j
+    match_matrix = objects.map obj->
+        criteria.map criterion->
+            name = name_mapper obj
+            here.match_single_criterion name criterion matching_strategy
+
+    # Checks if the ith object is matched by any criterion.
+    is_object_matched_by_anything : Integer -> Boolean
+    is_object_matched_by_anything i =
+        match_matrix.at i . any x->x
+
+    # Checks if the ith criterion matches any columns.
+    does_criterion_match_anything : Integer -> Boolean
+    does_criterion_match_anything i =
+        match_matrix.map (col -> col.at i) . any x->x
+
+    # Selects object indices which satisfy the provided predicate.
+    select_matching_indices : (Integer -> Boolean) -> Vector Text
+    select_matching_indices matcher =
+        0.up_to objects.length . to_vector . filter matcher
+
+    # Check consistency
+    case on_problems of
+        Ignore -> Nothing
+        Report_Warning ->
+            criteria.map_with_index j-> criterion->
+                if does_criterion_match_anything j then Nothing else
+                    error = No_Matches_Found criterion
+                    warning_callback error.to_display_text
```

Comment:
This seems useless – if I understand correctly, `warning_callback` is supposed to be replaced by the dataflow warnings system. But this result is dropped, so what's up.

Reply:
This was written before we saw any design for the warning system. The idea is that we will modify this accordingly depending on the actual warning implementation. For now this is just a placeholder ensuring that we can test and document that we are 'in some way' reporting warnings.